### PR TITLE
julia-bin: 1.8.5 -> 1.9.0

### PR DESCRIPTION
**Pull request by GaetanLepage**
###### Description of changes

Changelog: https://github.com/JuliaLang/julia/blob/v1.9.0/NEWS.md

###### Things done
cc @NickCao @joshniemela 

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by GaetanLepage** [comment]
@wegank regarding https://github.com/JuliaLang/julia/pull/47914, should I keep this patching strategy or try to use the `--skip internet_required` flag ?

**Comment by wegank** [comment]
> @wegank regarding [JuliaLang/julia#47914](https://github.com/JuliaLang/julia/pull/47914), should I keep this patching strategy or try to use the `--skip internet_required` flag ?

Wait, the patch still applies to 1.9.0?

**Comment by GaetanLepage** [comment]
> Wait, the patch still applies to 1.9.0?

I guess... Is there any reason we should not skipping internet depending tests now ?

**Comment by wegank** [comment]
> > Wait, the patch still applies to 1.9.0?
> 
> I guess... Is there any reason we should not be skipping internet depending tests now ?

Then the patch has to be kept so that the `internet_required` meta-tag exists...

**Comment by wegank** [comment]
Also, `jula-bin` -> `julia-bin`

**Comment by GaetanLepage** [comment]
> Then the patch has to be kept so that the `internet_required` meta-flag exists...

You are right. I was thinking that as the PR has been merged, the patch wasn't necessary anymore.
Turns out the 1.9.0 release has not received the change, so we need to keep our patch.

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 230965` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>julia-bin</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>julia_18-bin</li>
  </ul>
</details>

File: pkgs/development/compilers/julia/patches/1.9-bin/0001-allow-skipping-internet-required-tests.patch [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
++    "LibCURL",
++    "LibGit2",
++    "Pkg",
++    "download",
++    "TOML",
++]
++
++const NETWORK_REQUIRED_LIST = vcat(INTERNET_REQUIRED_LIST, ["Sockets"])
++
+ """
```
Comment (wegank) [inline pkgs/development/compilers/julia/patches/1.9-bin/0001-allow-skipping-internet-required-tests.patch:22]: Looks like `filtertests!(tests, "internet_required", INTERNET_REQUIRED_LIST)` has been dropped from the patch?

Comment (GaetanLepage) [reply pkgs/development/compilers/julia/patches/1.9-bin/0001-allow-skipping-internet-required-tests.patch:22]: Indeed, nice catch. I just added it back.

**Comment by GaetanLepage** [comment]
> Build log: https://gist.github.com/GaetanLepage/bdeb8792cbdb10660e408d943689813a

The `codegen` test still fails. It seems to complain about some missing library.
https://github.com/JuliaLang/julia/blob/8e630552924eac54c809aa7bc30871c7df1582d3/test/compiler/codegen.jl#L684-L708

**Comment by wegank** [comment]
Looks like `compiler/codegen` passed on the Darwin side...

**Comment by GaetanLepage** [comment]
> Looks like `compiler/codegen` passed on the Darwin side...

https://github.com/JuliaLang/julia/blob/v1.9.0/NEWS.md#external-dependencies

**Comment by wegank** [comment]
GitHub broken once a day 🤬

**Comment by GaetanLepage** [comment]
> https://github.com/JuliaLang/julia/blob/v1.9.0/NEWS.md#external-dependencies

@wegank, seems like we would have to link `libstdc++` ourselves, right ?

**Comment by GaetanLepage** [comment]
> Maybe not, the log looks like it's successfully loading some libraries when it's not supposed to success, maybe related to our RPATH patching.
> 
> Looking at https://github.com/JuliaLang/julia/blob/9002d1665cd849858fc3150361f5b33aacfaf688/test/compiler/codegen.jl#L688. The tests copies binaries into a temporary prefix, then remove the `libjulia-codegen.so` library to test the binaries behavior to see if it still works, however as the RPATHs were already patched to reference the installed libraries, `libjulia-codegen.so` could still be found and the expected failures just cannot be reproduced. I think this is a false positive, we may simply disable that particular test.

Indeed, skipping the test leads to a successful build. Do you see anything else to do ?

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 230965` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>julia-bin (julia-stable-bin ,julia_19-bin)</li>
  </ul>
</details>

**Comment by NickCao** [comment]
Everything looks fine now, just waiting for ofborg.

**Pull request merged**